ntoskrnl/io/iomgr/deviface.c

```diff
@@ -30,185 +31,437 @@ static PWCHAR BaseKeyString = L"\\Registry\\Machine\\System\\CurrentControlSet\\
 
 static
 NTSTATUS
-OpenRegistryHandlesFromSymbolicLink(IN PUNICODE_STRING SymbolicLinkName,
-                                    IN ACCESS_MASK DesiredAccess,
-                                    IN OPTIONAL PHANDLE GuidKey,
-                                    IN OPTIONAL PHANDLE DeviceKey,
-                                    IN OPTIONAL PHANDLE InstanceKey)
+IopBuildSymbolicLink(_In_ PUNICODE_STRING PrefixString,
+                     _In_ PUNICODE_STRING DeviceString,
+                     _In_ PUNICODE_STRING GuidString,
+                     _In_opt_ PUNICODE_STRING ReferenceString,
+                     _Out_ PUNICODE_STRING SymbolicLinkName)
 {
-    OBJECT_ATTRIBUTES ObjectAttributes;
-    UNICODE_STRING BaseKeyU;
-    UNICODE_STRING GuidString, SubKeyName, ReferenceString;
-    PWCHAR StartPosition, EndPosition;
-    HANDLE ClassesKey;
-    PHANDLE GuidKeyRealP, DeviceKeyRealP, InstanceKeyRealP;
-    HANDLE GuidKeyReal, DeviceKeyReal, InstanceKeyReal;
+    UNICODE_STRING SymbolicLink;
     NTSTATUS Status;
 
-    SubKeyName.Buffer = NULL;
+    /* Build up new symbolic link */
+    SymbolicLink.Length = 0;
+    SymbolicLink.MaximumLength = PrefixString->Length + DeviceString->Length + sizeof(L"#") + GuidString->Length + (ReferenceString != NULL ? ReferenceString->Length : sizeof(L"\\")) + sizeof(UNICODE_NULL);
```

Comment:
```suggestion
    SymbolicLink.MaximumLength =
        PrefixString->Length + DeviceString->Length
        + sizeof(L"#") + GuidString->Length
        + (ReferenceString != NULL ? ReferenceString->Length : sizeof(L"\\"))
        + sizeof(UNICODE_NULL);
```
Additional observations: both the `sizeof(L"#")` and `sizeof(L"\\")` count their own null-terminators, so in total you have got space for 2 more NULL terminators "for free" 😂

Reply:
By the way, in order to implement the `RtlUnicodeStringPrintf` suggestion from below, having
- in the declarations, a `UNICODE_STRING PathSep = RTL_CONSTANT_STRING(L"\\");` ,
- somewhere (possibly before that `SymbolicLink.MaximumLength` assignation) something like:
```c
    if (!ReferenceString)
        ReferenceString = &PathSep;
```
Then, in that `SymbolicLink.MaximumLength` from above, the line
```c
(ReferenceString != NULL ? ReferenceString->Length : sizeof(L"\\"))
```
could be replaced by a simple `ReferenceString->Length`

Reply:
> Additional observations: both the sizeof(L"#") and sizeof(L"\\") count their own null-terminators, so in total you have got space for 2 more NULL terminators "for free" :joy:

Ah, I know: to avoid this, I can simply subtract the length of a null terminator multiplied by `2`. It will remove additional unneeded length, but I doubt a bit: is it the good way to do that, or I can specify the length of these characters without null terminator somehow (probably via the fixed integer value)? :thinking:

Reply:
Ah sorry, now I understand! With your suggestion from the 2nd comment (and the one below), there are no the two unneeded null terminators anymore. So it is correct as-is now I guess. Thanks, fixed.
Fix me if I'm wrong and my previous comment is still actual.